sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -171,122 +171,52 @@ case class CreateMap(children: Seq[Expression]) extends Expression {
-
-    def defaultName: Literal = Literal( s"col${idx + 1}" )
-    def nameExpression( e : Expression ) : Literal = e match {
-      case ne : NamedExpression => Literal(ne.name)
-      case _ => defaultName
-    }
-  }
-
-  private def expressionInfo: ExpressionInfo = new ExpressionInfo( "CreateStruct",
-    "struct",
-    "_FUNC_(col1, col2, col3, ...)",
-    "Creates a struct with the given field values.")
-  def registryEntry: (String, (ExpressionInfo, CreateStruct.type)) =
-    "struct" -> (expressionInfo -> this )
-}
-
-/**
- * Creates a struct with the given field names and values
- *
- * @param children Seq(name1, val1, name2, val2, ...)
- */
 // scalastyle:off line.size.limit
 @ExpressionDescription(
-  usage = "_FUNC_(name1, val1, name2, val2, ...) - Creates a struct with the given field names and values.")
+usage = "_FUNC_(col1, col2, col3, ...) - Creates a struct with the given field values.")
 // scalastyle:on line.size.limit
-case class CreateNamedStruct(children: Seq[Expression]) extends Expression {
+case class CreateStruct(children: Seq[Expression]) extends Expression with Unevaluable {
+  override lazy val resolved: Boolean = false
+  override def nullable: Boolean = false
+  override def dataType: DataType = throw new UnresolvedException(this, "dataType")
+  override def prettyName: String = "struct"
+}
 
-  /**
-   * Returns Aliased [[Expression]]s that could be used to construct a flattened version of this
-   * StructType.
-   */
-  def flatten: Seq[NamedExpression] = valExprs.zip(names).map {
-    case (v, n) => Alias(v, n.toString)()
-  }
+/**
+ * Common base class for both [[CreateNamedStruct]] and [[CreateNamedStructUnsafe]].
+ */
+trait CreateNamedStructLike extends Expression {
+  lazy val (nameExprs, valExprs) = children.grouped(2).map {
+    case Seq(name, value) => (name, value)
+  }.toList.unzip
 
-  private lazy val (nameExprs, valExprs) =
-    children.grouped(2).map { case Seq(name, value) => (name, value) }.toList.unzip
+  lazy val names = nameExprs.map(_.eval(EmptyRow))
 
-  def valueExpressions: List[Expression] = valExprs
+  override def nullable: Boolean = false
 
-  private lazy val names = nameExprs.map(_.eval(EmptyRow))
+  override def foldable: Boolean = valExprs.forall(_.foldable)
 
   override lazy val dataType: StructType = {
     val fields = names.zip(valExprs).map {
-      case (name, valExpr: NamedExpression) =>
-        StructField(name.asInstanceOf[UTF8String].toString,
-          valExpr.dataType, valExpr.nullable, valExpr.metadata)
-      case (name, valExpr) =>
-        StructField(name.asInstanceOf[UTF8String].toString,
-          valExpr.dataType, valExpr.nullable, Metadata.empty)
+      case (name, expr) =>
+        val metadata = expr match {
+          case ne: NamedExpression => ne.metadata
+          case _ => Metadata.empty
+        }
+        StructField(name.toString, expr.dataType, expr.nullable, metadata)
```

Comment:
This was already a problem for CreateNamedStruct; we may need to address it. It should normally not be a problem as this is normally used in ML for adding metadata to top-level aliases in a project (which do not get erased).